feat: add formatted model files

KennetLindquist opened the pull request:
### Main improvements in this PR:
- Added model files: `json`, `mat`, `txt`, and `xlsx`.

**I hereby confirm that I have:**

- [x] Tested my code on my own computer for running the model

KennetLindquist committed:
feat: add exportFromModel.m function

feat: add exportFromModel.m function to convert XML or YML model files to MAT, TXT, XLSX, XML, and YML

Uses `exportForGit.m` function from the [RAVEN toolbox](https://github.com/SysBioChalmers/RAVEN/blob/main/io/exportForGit.m).

KennetLindquist committed:
feat: add exportToJson.py function

feat: add exportToJson.py function to convert XML model files to JSON

Uses the functions read_sbml_model and save_json_model from the [cobrapy](https://github.com/opencobra/cobrapy) package.

KennetLindquist committed:
feat: add exportAnnotationsToTsv.m function

feat: add exportAnnotationsToTsv.m function to export an XML model file to tab-delimited annotation files. Annotation files are separated into reactions, metabolites, genes, model, and compartments.

Uses exportToTabDelimited function from the [RAVEN toolbox](https://github.com/SysBioChalmers/RAVEN/blob/main/io/exportToTabDelimited.m).

KennetLindquist committed:
feat: add iML1515.yml

feat: add iML1515.yml using the exportFromModel.m function on iML1515.xml.

KennetLindquist committed:
feat: add genes.tvs, mets.tsv, and rxns.tsv

feat: add genes.tvs, mets.tsv, and rxns.tsv using the exportAnnotationsToTsv.m function on iML1515.xml.

KennetLindquist committed:
feat(memote): add YML file support and utility functions

feat(memote): add YML file support and utility functions.
- Update `memote_test.py` to enable YML file usage
- Introduce `memote_test_utils.py` with helper functions for Memote test YML file usage

mihai-sysbio committed:
chore: update branch references

KennetLindquist committed:
feat: format model files

### Main improvements in this PR:
- Added model export functions to the following formats: `mat`, `txt`, `xlsx`, `xml`, `yml`, and `json`.
- Added model annotation export function.
- Added the iML1515.yml model file.
- Added annotation TSV files: `genes.tsv`, `mets.tsv`, and `rxns.tsv`.
- Implemented `yml` file support in the Memote test workflow in preparation for removing the `xml` model file from the devel branch.

**I hereby confirm that I have:**

- [x] Tested my code on my own computer for running the model

KennetLindquist committed:
chore(model): rename iML1515 to Ecoli-GEM

chore(model): rename iML1515 to Ecoli-GEM including instances inside model

KennetLindquist committed:
feat: add formatted model files

feat: add formatted model files `.json`, `.mat`, `.txt`, and `.xlsx`.

KennetLindquist committed:
ci(memote): add Memote HTML report [skip ci]

cdebuck commented on code/memote_test_utils.py:
```diff
@@ -0,0 +1,115 @@
+# Function to append annotation information to the model
+def append_annotations(model, genes_df, mets_df, rxns_df):
+    """Adds annotation data to the model from TSV files."""
+    # Ensure all genes have SBO:0000243
```
It might be good to be explicit about what this means and why it is done (so the standard annotation for a gene that every gene is expected to have as default.

KennetLindquist replied:
Excellent point! I have edited it to clarify.

cdebuck commented on code/memote_test_utils.py:
```diff
@@ -0,0 +1,115 @@
+    """Adds annotation data to the model from TSV files."""
+    # Ensure all genes have SBO:0000243
+    for gene in model.genes:
+        gene.annotation.setdefault("sbo", "SBO:0000243")
```
I assume this does not override existing annotations? When the default notation in the .xml file would be something else, is that information now lost perhaps?
Edit

KennetLindquist replied:
This does potentially overwrite existing annotations. It would be better practice to implement a function that fetches the sbo term and updates it, rather than overwriting it. However, this workflow will shortly become deprecated, as the memote report will be exclusively generated from the xml model file. It is therefore deemed unnecessary to implement this change at this time.

cdebuck commented on code/memote_test_utils.py:
```diff
@@ -0,0 +1,115 @@
+    # Update gene annotations
+    for _, row in genes_df.iterrows():
+        gene_id_clean = row[genes_df.columns[0]].removeprefix("G_")  # Cleaned ID
+        if gene_id_clean in model.genes:
```
Do we want to raise an exception if the gene_id_clean is not present in the model?

Why do we prefer to run through the id's via the tsv files instead of the model-id's?

KennetLindquist replied:
It would be an improvement raise an exeption or have the code hard-fail in that case. We run through the id's via the tsv files out of convenience, not preference. Because this workflow will shortly become deprecated it is unnecessary to implement any changes at this time.

cdebuck commented on code/memote_test_utils.py:
```diff
@@ -0,0 +1,115 @@
+                annotations["sbo"] = sbo_term
+            
+            rxn.annotation.update(annotations)
+            
```
What happens to the model object after this function? I don't see an return statement.

It also feels a bit strange to add annotation information to a model object from a file that has been generated based on the model object.
I assume this is due to the yml/xml file conversion issues, but there my suggestion would be to always work with the model object generated directly from the .xml file which should not be missing any info.

KennetLindquist replied:
This function directly updates the model, so it is not necessary to have a return statement. However, it would be clearer with a return statement and would potentially increase the usability of this function.

You are correct in your assumption about the yml/xml conversion issue. This is why this workflow was implemented in the first place. 

Because this workflow will shortly become deprecated it is unnecessary to implement any changes at this time.

cdebuck commented on code/memote_test_utils.py:
```diff
@@ -0,0 +1,115 @@
+    
+            # Update EC-NUMBER key (required by MEMOTE)
+            if "EC-NUMBER" in annotations:
+                annotations["ec-code"] = annotations.pop("EC-NUMBER")
```
Does this give any issue with the element that annotation.pop("EC-NUMBER") returns? I can imagine that is a pandas dataframe object, I think that might not be the type of object that can be linked to an reaction-annotation

cdebuck replied:
Also: could there be a more elegant way for renaming a column in the annotations dataframe (https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.rename.html), because that is what is happening in line 100 if I understand it correctly

KennetLindquist replied:
You are deffinitely right that renaming would be more elegant if it is a dataframe object. However, because this workflow will shortly become deprecated it is unnecessary to implement any changes at this time.

KennetLindquist committed:
docs: add detail to comment for clarity

KennetLindquist committed:
revert: remove genes.tsv, mets.tsv, and rxns.tsv annotation files

revert: remove genes.tsv, mets.tsv, and rxns.tsv annotation files. These annotation files structure and content do not suit their intended task of supplementing the yml model file with missing annotation information. These are removed for now and will be added back in a later release.

mihai-sysbio commented:
It appears there are a few instances of iML1515 remaining, let's look at those.

mihai-sysbio commented:
> It appears there are a few instances of iML1515 remaining, let's look at those.

This has been resolved on `develop`.
